File: backend/lib/frontmatter.py
import re


def parse_frontmatter(content: str) -> dict:
    if not content.startswith("---"):
        return {"frontmatter": {}, "body": content}

    end_index = content.find("\n---", 3)
    if end_index == -1:
        return {"frontmatter": {}, "body": content}

    fm_raw = content[4:end_index].strip()
    body = content[end_index + 4:]
    if body.startswith("\n"):
        body = body[1:]

    frontmatter: dict = {}

    for line in fm_raw.split("\n"):
        match = re.match(r"^(\w+):\s*(.*)$", line)
        if not match:
            continue
        key, value = match.group(1), match.group(2)

        if key == "tags":
            cleaned = value.strip().strip("[]").strip()
            if cleaned:
                frontmatter["tags"] = [t.strip() for t in cleaned.split(",") if t.strip()]
            else:
                frontmatter["tags"] = []
        else:
            frontmatter[key] = value

    return {"frontmatter": frontmatter, "body": body}


def serialize_frontmatter(frontmatter: dict, body: str) -> str:
    entries: list[str] = []

    if frontmatter.get("tags") and len(frontmatter["tags"]) > 0:
        entries.append(f"tags: [{', '.join(frontmatter['tags'])}]")

    for key, value in frontmatter.items():
        if key == "tags":
            continue
        if value is not None:
            entries.append(f"{key}: {value}")

    if not entries:
        return body

    return f"---\n{chr(10).join(entries)}\n---\n{body}"


def update_tags(content: str, tags: list[str]) -> str:
    parsed = parse_frontmatter(content)
    fm = parsed["frontmatter"]
    fm["tags"] = tags
    return serialize_frontmatter(fm, parsed["body"])

File: lib/frontmatter.ts
/**
 * 简易 frontmatter 解析器
 * 支持 YAML-like 格式的 tags 字段
 */

export interface Frontmatter {
  tags?: string[];
  [key: string]: unknown;
}

interface ParsedContent {
  frontmatter: Frontmatter;
  body: string;
}

/**
 * 解析 markdown 内容中的 frontmatter
 */
export function parseFrontmatter(content: string): ParsedContent {
  if (!content.startsWith('---')) {
    return { frontmatter: {}, body: content };
  }

  const endIndex = content.indexOf('\n---', 3);
  if (endIndex === -1) {
    return { frontmatter: {}, body: content };
  }

  const fmRaw = content.slice(4, endIndex).trim();
  const body = content.slice(endIndex + 4).replace(/^\n/, '');
  const frontmatter: Frontmatter = {};

  for (const line of fmRaw.split('\n')) {
    const match = line.match(/^(\w+):\s*(.*)$/);
    if (!match) continue;
    const [, key, value] = match;

    if (key === 'tags') {
      // 支持 [tag1, tag2] 和 tag1, tag2 格式
      const cleaned = value.replace(/^\[|\]$/g, '').trim();
      if (cleaned) {
        frontmatter.tags = cleaned.split(',').map((t) => t.trim()).filter(Boolean);
      } else {
        frontmatter.tags = [];
      }
    } else {
      frontmatter[key] = value;
    }
  }

  return { frontmatter, body };
}

/**
 * 将 frontmatter 和 body 重新序列化
 */
export function serializeFrontmatter(frontmatter: Frontmatter, body: string): string {
  const entries: string[] = [];

  if (frontmatter.tags && frontmatter.tags.length > 0) {
    entries.push(`tags: [${frontmatter.tags.join(', ')}]`);
  }

  for (const [key, value] of Object.entries(frontmatter)) {
    if (key === 'tags') continue;
    if (value !== undefined && value !== null) {
      entries.push(`${key}: ${String(value)}`);
    }
  }

  if (entries.length === 0) return body;

  return `---\n${entries.join('\n')}\n---\n${body}`;
}

